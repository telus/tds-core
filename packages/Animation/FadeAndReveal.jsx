import React from 'react'
import PropTypes from 'prop-types'
import { Transition } from 'react-transition-group'

const defaultStyle = (duration, timeout) => ({
  transition: `height ${duration || timeout}ms, opacity ${duration || timeout}ms ease-in-out`,
  opacity: 0,
  height: 0,
  overflow: 'hidden',
})

const transitionStyles = height => (duration, timeout) => ({
  entering: { opacity: 1, height },
  entered: { opacity: 1, height: 'auto' },
  exiting: {
    opacity: 0,
    height,
    transition: `height ${duration || timeout}ms, opacity ${duration || timeout}ms ease-in-out`,
  },
  exited: { opacity: 0 },
})

const FadeAndReveal = ({ delay, height, children, ...rest }) => (
  <Transition {...rest}>
    {state => (
      <div
        style={{
          ...defaultStyle(rest.duration, rest.timeout),
          ...transitionStyles(height)(rest.duration, rest.timeout)[state],
        }}
      >
        {children()}
      </div>
    )}
  </Transition>
)

FadeAndReveal.propTypes = {
  height: PropTypes.number.isRequired,
  duration: PropTypes.number.isRequired,
  delay: PropTypes.number,
  children: PropTypes.func.isRequired,
}

FadeAndReveal.defaultProps = {
  delay: 0,
}

export default FadeAndReveal
