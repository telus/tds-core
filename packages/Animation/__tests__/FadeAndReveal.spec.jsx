import React from 'react'
import { render, mount } from 'enzyme'

import FadeAndReveal, { StyledContainer } from '../FadeAndReveal'

describe('FadeAndReveal', () => {
  it('renders when the transition is triggered', () => {
    const fadeAndReveal = render(
      <FadeAndReveal timeout={500} duration={100} height={500} in={true}>
        {() => <div>Content to fade</div>}
      </FadeAndReveal>
    )

    expect(fadeAndReveal).toMatchSnapshot()
  })

  it('renders when the transition is not triggered', () => {
    const fadeAndReveal = render(
      <FadeAndReveal timeout={500} duration={100} height={500} in={false}>
        {() => <div>Content to fade</div>}
      </FadeAndReveal>
    )

    expect(fadeAndReveal).toMatchSnapshot()
  })

  it('should hide content from screenreaders when exited', () => {
    const fadeAndReveal = mount(
      <FadeAndReveal timeout={500} duration={100} height={500} in={false}>
        {() => <div>Content to fade</div>}
      </FadeAndReveal>
    )

    const styledContainer = fadeAndReveal.find(StyledContainer)
    expect(styledContainer).toHaveProp('aria-hidden', true)
    expect(styledContainer.prop('style')).toHaveProperty('visibility', 'hidden')
  })

  it('should display content from screenreaders when entered', () => {
    const fadeAndReveal = mount(
      <FadeAndReveal timeout={500} duration={100} height={500} in={true}>
        {() => <div>Content to fade</div>}
      </FadeAndReveal>
    )

    const styledContainer = fadeAndReveal.find(StyledContainer)
    expect(styledContainer).toHaveProp('aria-hidden', false)
    expect(styledContainer.prop('style')).toHaveProperty('visibility', 'visible')
  })
})
