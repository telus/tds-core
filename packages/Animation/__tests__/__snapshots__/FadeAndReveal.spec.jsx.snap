// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FadeAndReveal renders when the transition is not triggered 1`] = `
@media (prefers-reduced-motion:reduce) {
  .c0 {
    -webkit-transition: none !important;
    transition: none !important;
  }
}

<div
  aria-hidden="true"
  class="c0"
  style="opacity:0;height:0px;overflow:hidden;visibility:hidden;transition:height 500ms ease-in-out, opacity 500ms ease-in-out, visibility 0ms 500ms"
>
  <div>
    Content to fade
  </div>
</div>
`;

exports[`FadeAndReveal renders when the transition is triggered 1`] = `
@media (prefers-reduced-motion:reduce) {
  .c0 {
    -webkit-transition: none !important;
    transition: none !important;
  }
}

<div
  aria-hidden="false"
  class="c0"
  style="opacity:1;height:auto;overflow:hidden;visibility:visible;transition:unset"
>
  <div>
    Content to fade
  </div>
</div>
`;
